WordPress/Classes/System/WordPressAppDelegate.m

```diff
@@ -518,8 +549,63 @@ - (void)showWelcomeScreenIfNeededAnimated:(BOOL)animated
     }
 }
 
+
+- (void)showSigninScreenIfNeededAnimated:(BOOL)animated
+{
+    if (self.isWelcomeScreenVisible || !([self noSelfHostedBlogs] && [self noWordPressDotComAccount])) {
+        return;
+    }
+
+    UIViewController *presenter = self.window.rootViewController;
+    // Check if the presentedVC is UIAlertController because in iPad we show a Sign-out button in UIActionSheet
+    // and it's not dismissed before the check and `dismissViewControllerAnimated` does not work for it
+    if (presenter.presentedViewController && ![presenter.presentedViewController isKindOfClass:[UIAlertController class]]) {
+        [presenter dismissViewControllerAnimated:animated completion:^{
+            [self showSigninScreenAnimated:animated thenEditor:NO];
+        }];
+    } else {
+        [self showSigninScreenAnimated:animated thenEditor:NO];
+    }
+}
+
+- (void)showSigninScreenAnimated:(BOOL)animated thenEditor:(BOOL)thenEditor
+{
+    [self showSigninScreen:nil animated:animated thenEditor:thenEditor];
+}
+
+- (void)showSigninScreen:(NSString *)token animated:(BOOL)animated thenEditor:(BOOL)thenEditor
+{
+    // Complete a magic link signin if needed.
+    NSString *savedEmail = [SigninHelpers getEmailAddressForTokenAuth];
+    if (token && savedEmail) {
+        SigninLinkAuthViewController *controller = [SigninLinkAuthViewController controller:savedEmail token:token];
+        UINavigationController *navigationController = [[NUXNavigationController alloc] initWithRootViewController:controller];
+        UIViewController *rootViewController = self.window.rootViewController;
+        if (rootViewController.presentedViewController)  {
+            [rootViewController.presentedViewController dismissViewControllerAnimated:NO completion:^{
```

Comment:
Why do we need dismissing here? I fear there's a chance this could dismiss an unsaved post or comment reply. I haven't been able to test the magic link flow, and the new sign in only seems to be enabled for now for the first time setup. But I would expect this could make its way into all the others sign in flows, so my worry is that the following could happen:

1. You have a working self hosted blog in the app already
2. Try to log in to wp.com using the email link.
3. Email is a bit slow today so you give up for now and go back to replying to comments on your self hosted.
4. A notification pops up with new mail
5. Go to mail, tap the link, this is called, and your half written comment is lost

Another potential problematic flow:

1. You install the app and try to log in to wp.com using the magic link
2. Email is not arriving, so you decide to use the password instead
3. You start writing a post
4. The email finally arrives, and for some reason you tap the link
5. This is called, the editor is dismiss (I'd hope the content isn't lost though)
6. Not sure what else happens, but the rest looks safe